sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java

```diff
@@ -95,4 +107,54 @@ public void counterToCell() {
     counter.dec();
     assertThat(cell.getCumulative(), CoreMatchers.equalTo(42L));
   }
+
+  @Category({RunnableOnService.class, UsesMetrics.class})
+  @Test
+  public void metricsE2E() {
+    final Counter count = Metrics.counter(MetricsTest.class, "count");
+    Pipeline pipeline = TestPipeline.create();
+    pipeline
+        .apply(Create.of(5, 8, 13))
+        .apply("MyStep1", ParDo.of(new DoFn<Integer, Integer>() {
+          @ProcessElement
+          public void processElement(ProcessContext c) {
+            Distribution values = Metrics.distribution(MetricsTest.class, "input");
+            count.inc();
+            values.update(c.element());
+
+            c.output(c.element());
+            c.output(c.element());
+          }
+        }))
+        .apply("MyStep2", ParDo.of(new DoFn<Integer, Integer>() {
+          @ProcessElement
+          public void processElement(ProcessContext c) {
+            Distribution values = Metrics.distribution(MetricsTest.class, "input");
+            count.inc();
+            values.update(c.element());
+          }
+        }));
+    PipelineResult result = pipeline.run();
+
+    result.waitUntilFinish();
+
+    MetricQueryResults metrics = result.metrics().queryMetrics(MetricsFilter.builder()
+      .addNameFilter(MetricNameFilter.inNamespace(MetricsTest.class))
+      .build());
+    final String step1Name = "MyStep1/AnonymousParDo/AnonymousParMultiDo";
+    assertThat(metrics.counters(), hasItem(
+        metricResult(MetricsTest.class.getName(), "count", step1Name, 3L, 3L)));
+    assertThat(metrics.distributions(), hasItem(
+        metricResult(MetricsTest.class.getName(), "input", step1Name,
+            DistributionResult.create(26L, 3L, 5L, 13L),
+            DistributionResult.create(26L, 3L, 5L, 13L))));
+
+    final String step2Name = "MyStep2/AnonymousParDo/AnonymousParMultiDo";
+    assertThat(metrics.counters(), hasItem(
+        metricResult(MetricsTest.class.getName(), "count", step2Name, 6L, 6L)));
```

Comment:
I'm worried about including physical values in the matcher across all runners;

The physical value should generally be `atLeast(6L)`; TODO and JIRA is acceptable rather than refactoring all of the `MetricMatchers`, but this should be something we're ready to fix.